File: backend/internal/dtos/Websocket.go
package dtos

import "kahoot-api/internal/models"

type MessageType = uint8

const (
	SCORES MessageType = iota
	QUESTION
	REGISTER
	ANSWER
)

type (
	QuestionMessage struct {
		Type        MessageType `json:"type"`
		GameId      string      `json:"game_id"`
		Title       string      `json:"title"`
		Body        string      `json:"body"`
		OptionOne   string      `json:"option_one"`
		OptionTwo   string      `json:"option_two"`
		OptionThree string      `json:"option_three"`
		OptionFour  string      `json:"option_four"`
		Correct     string      `json:"correct"`
	}

	ScoresMessage struct {
		Type    MessageType     `json:"type"`
		GameId  string          `json:"game_id"`
		Players []models.Player `json:"players"`
	}

	RegisterRequest struct {
		Type MessageType `json:"type"`
		Name string      `json:"name"`
	}

	AnswerRequest struct {
		Type   MessageType `json:"type"`
		Answer string      `json:"answer"`
	}
)

File: src/types/websocket.ts
import { QuizStatus, WebsocketMessage } from "./quiz";

export type WebsocketErrorMessage = {
  type: WebsocketMessage.ERROR;
  message: string;
};

type Player = {
  id: string;
  game_id: string;
  name: string;
  score: number;
};

export type WebsocketGameStateMessage = {
  type: WebsocketMessage.GAME_STATE;
  title: string;
  code: string;
  current_question: number;
  question_count: number;
  max_players: number;
  active_players: number;
  status: QuizStatus;
  players: Player[];
};

export type WebsocketQuestionMessage = {
  type: WebsocketMessage.QUESTION;
  gameid: string;
  title: string;
  body: string;
  option_one: string;
  option_two: string;
  option_three: string;
  option_four: string;
  correct: string;
};

export type WebsocketScoresMessage = {
  type: WebsocketMessage.SCORES;
  game_id: string;
  players: Player[];
};
